firmware/hw_layer/digital_input/digital_input_exti.cpp

```diff
@@ -103,9 +116,88 @@ digital_input_s* startDigitalCapture(const char *msg, brain_pin_e brainPin) {
 }
 #endif // EFI_ICU_INPUTS
 
-#endif /* HAL_USE_PAL && EFI_PROD_CODE */
+static inline void triggerInterrupt() {
+	// Manually fire the I2C1_EV interrupt, it will be queued after this interrupt returns
+	NVIC->STIR = I2C1_EV_IRQn;
+}
+
+CH_IRQ_HANDLER(STM32_I2C1_EVENT_HANDLER) {
+	OSAL_IRQ_PROLOGUE();
+
+	for (size_t i = 0; i < 16; i++) {
+		auto& channel = channels[i];
+
+		if (channel.Timestamp != 0 && channel.Callback) {
+			channel.Callback(channel.CallbackData, channel.Timestamp);
+			channel.Timestamp = 0;
```

Comment:
not safe if an event on the same channel comes in at the same time.

Now you're making me want to make timestamps 32-bit, though then they couldn't represent 0 anyways.
I think it's safer if you have a separate byte array (or byte field) representing which channels are active.  It's safer than this code, without being as complicated as using a uint16_t bit field and ldrex/strex to manipulate it.

Reply:
Well, it's at least safe now, if not particularly pretty.  We're locked for only 4 instructions.

```
cpsid i
ldrd r2, r3, [r4, #16]      // load timestamp in to r2, r3
strd r6, r7, [r4, #16]      // zero timestamp field
cpsie i

```